01-data-structures/03-linked-lists/linked_list.rb

```diff
@@ -3,121 +3,119 @@
+      @tail.next = nil
+      @length -= 1
     else
       @head = nil
       @tail = nil
+      @length = 0
     end
   end
 
-  # This method prints out a representation of the list.
+
   def print
-    temp = @head
-    temp_next = @head.next
-    puts temp.data
-    puts temp_next.data
-    while temp_next.next
-      temp = temp.next
-      temp_next = temp.next
-      puts temp_next.data
-    end
+    find_node_prior(@tail, true) 
   end
 
-  # This method removes `node` from the list and must keep the rest of the list intact.
   def delete(node)
-    if @head == node
-      if @head.next == nil
-        @tail = nil
-      end
-      @head = @head.next
+    if @head != node
+      current_node = find_node_prior(node, false)      
+      delete_next_node(current_node)
     else
-      current_node = @head
-      next_node = @head.next
-      while next_node != node
-        current_node = next_node
-        next_node = current_node.next 
-      end
-      delete_next_node(current_node, next_node)
+      self.remove_front
     end
   end
 
-  # This method adds `node` to the front of the list and must set the list's head to `node`.
   def add_to_front(node)
     if @head
-      old_head = @head
+      node.next = @head
       @head = node
-      @head.next = old_head
+      @length += 1
     else
       @head = node
       @tail = @head
+      @length = 1
     end
   end
 
-  # This method removes and returns the first node in the Linked List and must set Linked List's head to the second node.
   def remove_front
     if @head
       @head = @head.next
       if !@head
         @tail = @head
       end
+      @length -= 1
     end
   end
-  
-  # added this due to the benchmark portion of the assignment asking for me to delete the nth element in the List.
-  # In order to do a like for like with an Array, thought appropriate to add this method.
+
   def delete_node_by_index(index)
-    if index == 0
-      if @head.next != nil
-        @head = @head.next
-      else
-        @tail = nil
-        @head = nil
-      end
+    if index != 0
+      previous_node = locate_node_by_index(index - 1)
+      delete_next_node(previous_node)
     else
-      current_node = locate_node_by_index(index - 1)
-      next_node = current_node.next
-      delete_next_node(current_node, next_node)
+      self.remove_front
     end
   end
 
   def  locate_node_by_index(index)
-    current_node = @head
-    next_node = @head.next
-    for i in (1..index)
-      current_node = next_node
-      next_node = current_node.next 
+    if index < @length and index >= 0
+      current_node = @head
+      for i in (1..index)
+        current_node = current_node.next
+      end
+      return current_node
+    else
+      return nil
+    end
+
+  end
+
+  def find_node_prior(said_node, print)
+    node = @head
+    if print
+      puts node.data
+      puts node.next.data
+    end
+    while node.next != said_node && node != @tail
+      node = node.next
+      if print
```

Comment:
I think you're talking about the `if print` and the `print` in general?
I just saw that to printing a linked list was essentially walking through the list itself from beginning to end. I saw a lot of similarity in the code between what was needed to print and `locate_node_by_index`. The only thing missing from said method was the actual printing. Thought it would be interesting.

I can change this.